src/CircularArrays.jl

```diff
@@ -27,20 +27,30 @@ Alias for [`CircularArray{T,1,A}`](@ref).
 """
 const CircularVector{T} = CircularArray{T, 1}
 
-@inline clamp_bounds(arr::CircularArray, I::Tuple{Vararg{Int}})::AbstractArray{Int, 1} = map(Base.splat(mod), zip(I, axes(arr.data)))
-
 CircularArray(data::AbstractArray{T,N}) where {T,N} = CircularArray{T,N}(data)
 CircularArray{T}(data::AbstractArray{T,N}) where {T,N} = CircularArray{T,N}(data)
 CircularArray(def::T, size) where T = CircularArray(fill(def, size))
 
-@inline Base.getindex(arr::CircularArray, i::Int) = @inbounds getindex(arr.data, mod(i, Base.axes1(arr.data)))
-@inline Base.setindex!(arr::CircularArray, v, i::Int) = @inbounds setindex!(arr.data, v, mod(i, Base.axes1(arr.data)))
-@inline Base.getindex(arr::CircularArray, I::Vararg{Int}) = @inbounds getindex(arr.data, clamp_bounds(arr, I)...)
-@inline Base.setindex!(arr::CircularArray, v, I::Vararg{Int}) = @inbounds setindex!(arr.data, v, clamp_bounds(arr, I)...)
+@inline Base.getindex(arr::CircularArray, i::Integer) =
+    @inbounds getindex(arr.data, mod(i, eachindex(LinearIndices(arr.data))))
+@inline Base.getindex(arr::CircularArray{T,N}, I::Vararg{<:Integer,N}) where {T,N} =
+    @inbounds getindex(arr.data, mod.(I, axes(arr))...)
+
+@inline Base.setindex!(arr::CircularArray, v, i::Integer) =
+    @inbounds setindex!(arr.data, v, mod(i, eachindex(LinearIndices(arr.data))))
+@inline Base.setindex!(arr::CircularArray{T,N}, v, I::Vararg{<:Integer,N}) where {T,N} =
+    @inbounds setindex!(arr.data, v, mod.(I, axes(arr))...)
+
 @inline Base.size(arr::CircularArray) = size(arr.data)
 @inline Base.axes(arr::CircularArray) = axes(arr.data)
+Base.parent(arr::CircularArray) = arr.data
 
-@inline Base.checkbounds(::CircularArray, _...) = nothing
+@inline function Base.checkbounds(arr::CircularArray, I...)
+    J = Base.to_indices(arr, I)
+    length(J) == 1 && return nothing
+    length(J) >= ndims(arr) && return nothing
+    throw(BoundsError(arr, I))
+end
```

Comment:
I believe this could probably be cleaned up, and be made more consistent with base array bounds checks (also error when dimensions are too high). Same speed as your implementation.
```suggestion
@inline function Base.checkbounds(arr::CircularArray, I...)
    J = length(Base.to_indices(arr, I))
    (J == 1 || J == ndims(arr)) && return nothing
    Base.throw_boundserror(arr, I)
end
```

Reply:
Base does allow more indices: `rand(2,2)[2,2,1,1,1]` is OK, and `size(rand(2,2), 5) == 1`. More surprisingly, `rand(2,2)[2,2,1,1,:]` is also allowed.

Reply:
Well, you learn something new every day. Makes sense to keep it like that then, since it's implicitly of size 1 in every higher dimension

Reply:
Another case is:
```
julia> const newaxis = [CartesianIndex()];

julia> a1[:,newaxis,:]
3×1×4 CircularArray(::Array{Int64, 3}):
[:, :, 1] =
  2
 10
 18
...

julia> Base.to_indices(a1, (:,newaxis,:)) |> length
3
```